File: aggregator.js
import fs from 'fs';
import path from 'path';
import os from 'os';
import { Op } from 'sequelize';
import sequelize from './db.js';
import { FileCache, Account } from './models.js';
import { getStorageQuota, listFiles, uploadFile, fetchAllFiles, getDriveService, deleteFile } from './driveService.js';

/**
 * Helper to extract a chunk of a specific size from a stream and write to a temporary file.
 * Automatically handles stream pausing and unshifting the remaining chunk back.
 */
async function extractStreamPart(fileStream, size, tempPath) {
  const writeStream = fs.createWriteStream(tempPath);
  let bytesWritten = 0;

  return new Promise((resolve, reject) => {
    function onData(chunk) {
      const remaining = size - bytesWritten;
      if (chunk.length <= remaining) {
        bytesWritten += chunk.length;
        writeStream.write(chunk);
        if (bytesWritten === size) {
          cleanup();
          resolve(bytesWritten);
        }
      } else {
        const part = chunk.slice(0, remaining);
        bytesWritten += part.length;
        writeStream.write(part);

        // Put the leftover back to the beginning of the stream
        fileStream.unshift(chunk.slice(remaining));
        cleanup();
        resolve(bytesWritten);
      }
    }

    function onEnd() {
      cleanup();
      resolve(bytesWritten);
    }

    function onError(err) {
      cleanup();
      reject(err);
    }

    function cleanup() {
      fileStream.removeListener('data', onData);
      fileStream.removeListener('end', onEnd);
      fileStream.removeListener('error', onError);
      writeStream.end();
    }

    fileStream.on('data', onData);
    fileStream.on('end', onEnd);
    fileStream.on('error', onError);
  });
}

/**
 * Get total merged storage quota of all connected accounts.
 */
export async function getTotalQuota(accounts) {
  let total = 0;
  let used = 0;
  const accountQuotas = [];

  for (const account of accounts) {
    try {
      const quota = await getStorageQuota(account);
      total += quota.total;
      used += quota.used;

      accountQuotas.push({
        account_id: account.id,
        email: quota.email,
        display_name: quota.display_name,
        total: quota.total,
        used: quota.used,
        free: quota.total - quota.used
      });

      // Update quota in database
      account.quota_total = quota.total;
      account.quota_used = quota.used;
      if (quota.display_name) {
        account.display_name = quota.display_name;
      }
      await account.save();
    } catch (e) {
      console.error(`Failed to fetch quota for account ${account.email}:`, e);
      accountQuotas.push({
        account_id: account.id,
        email: account.email,
        display_name: account.display_name || account.email,
        total: account.quota_total || 0,
        used: account.quota_used || 0,
        free: (account.quota_total || 0) - (account.quota_used || 0),
        error: e.message
      });
    }
  }

  return {
    total,
    used,
    free: total - used,
    accounts: accountQuotas
  };
}

/**
 * Fetch and merge lists of files from all accounts into one single virtual list.
 */
export async function getMergedFiles(accounts, folderId = 'root') {
  const merged = [];
  for (const account of accounts) {
    try {
      const files = await listFiles(account, folderId);
      merged.push(...files);
    } catch (e) {
      console.error(`Error listing files for account ${account.email}:`, e);
    }
  }

  // Sort: Folders first, then alphabetically by name
  merged.sort((a, b) => {
    const aIsDir = a.mimeType === 'application/vnd.google-apps.folder' ? 0 : 1;
    const bIsDir = b.mimeType === 'application/vnd.google-apps.folder' ? 0 : 1;
    if (aIsDir !== bIsDir) return aIsDir - bIsDir;
    return (a.name || '').toLowerCase().localeCompare((b.name || '').toLowerCase());
  });

  return merged;
}

/**
 * Groups split parts (*.gpart.NNN) into a single virtual merged file representation.
 */
export function mergeGpartFiles(filesList) {
  const pattern = /^(.*)\.gpart\.(\d+)$/;
  const groups = {}; // key: JSON.stringify([parent_id, base_name]) -> Array of parts
  const nonGparts = [];

  for (const f of filesList) {
    const match = f.name.match(pattern);
    if (match) {
      const baseName = match[1];
      const partNum = parseInt(match[2], 10);
      const parentId = f.parent_id;
      const key = JSON.stringify([parentId, baseName]);
      if (!groups[key]) {
        groups[key] = [];
      }
      groups[key].push({ partNum, file: f });
    } else {
      nonGparts.push(f);
    }
  }

  for (const key in groups) {
    const [parentId, baseName] = JSON.parse(key);
    const parts = groups[key];
    parts.sort((a, b) => a.partNum - b.partNum);

    const totalSize = parts.reduce((sum, p) => sum + (p.file.size || 0), 0);
    const modifiedTimes = parts.map(p => p.file.modifiedTime).filter(Boolean);
    const latestModified = modifiedTimes.length > 0 
      ? modifiedTimes.reduce((max, t) => t > max ? t : max, modifiedTimes[0])
      : null;

    const isStarred = parts.some(p => p.file.is_starred);
    const isShared = parts.some(p => p.file.is_shared);

    const firstPart = parts[0].file;
    const virtualId = `gpart:${parentId}:${baseName}`;

    nonGparts.push({
      id: virtualId,
      name: baseName,
      mimeType: firstPart.mimeType || 'application/octet-stream',
      size: parseInt(totalSize, 10),
      modifiedTime: latestModified || firstPart.modifiedTime,
      parent_id: parentId,
      account_id: firstPart.account_id,
      is_starred: isStarred,
      is_shared: isShared,
      is_gpart: true,
      parts: parts.map(p => p.file)
    });
  }

  return nonGparts;
}

/**
 * Upload files smartly into the account with the most free space.
 * If no single account fits the file, it splits it across multiple accounts.
 */
export async function smartUpload(accounts, fileStream, filename, mimeType, fileSize, folderId = 'root', progressCallback = null) {
  const accountSpaces = [];
  
  for (const account of accounts) {
    try {
      const quota = await getStorageQuota(account);
      const freeSpace = quota.total - quota.used;
      accountSpaces.push({ account, freeSpace, email: quota.email });
    } catch (e) {
      console.error(`Error checking quota for smartUpload on ${account.email}:`, e);
    }
  }

  // Sort by free space descending
  accountSpaces.sort((a, b) => b.freeSpace - a.freeSpace);

  // 1. Try uploading to a single account that fits the file
  for (const { account, freeSpace, email } of accountSpaces) {
    if (freeSpace >= fileSize) {
      const result = await uploadFile(account, fileStream, filename, mimeType, folderId, fileSize, progressCallback);
      return {
        success: true,
        file: result,
        uploaded_to: email,
        account_id: account.id
      };
    }
  }

  // 2. If no single account fits, split and upload across accounts (storage aggregation)
  const bufferSize = 10 * 1024 * 1024; // 10MB safe buffer per account
  const usableSpaces = [];
  let totalUsable = 0;

  for (const { account, freeSpace, email } of accountSpaces) {
    const usable = Math.max(0, freeSpace - bufferSize);
    if (usable > 0) {
      usableSpaces.push({ account, usable, email });
      totalUsable += usable;
    }
  }

  if (totalUsable < fileSize) {
    return {
      success: false,
      error: 'Tidak ada akun dengan ruang penyimpanan yang cukup untuk file ini.'
    };
  }

  // Allocate part sizes
  let remainingSize = fileSize;
  const allocations = [];
  for (const { account, usable, email } of usableSpaces) {
    if (remainingSize <= 0) break;
    const allocated = Math.min(remainingSize, usable);
    allocations.push({ account, allocatedSize: allocated, email });
    remainingSize -= allocated;
  }

  const uploadedParts = [];
  try {
    for (let i = 0; i < allocations.length; i++) {
      const { account, allocatedSize, email } = allocations[i];
      const partNum = i + 1;
      const partFilename = `${filename}.gpart.${String(partNum).padStart(3, '0')}`;

      // Create temporary file to hold the chunk
      const tempPath = path.join(os.tmpdir(), `gpart_chunk_${Date.now()}_part_${partNum}`);
      const written = await extractStreamPart(fileStream, allocatedSize, tempPath);

      if (written < allocatedSize) {
        throw new Error(`Unexpected EOF: expected ${allocatedSize} bytes, got ${written}`);
      }

      // Upload chunk to Drive
      const partStream = fs.createReadStream(tempPath);
      const partResult = await uploadFile(account, partStream, partFilename, mimeType, folderId, written, progressCallback);

      uploadedParts.push({
        account,
        file: partResult,
        name: partFilename,
        uploaded_to: email,
        account_id: account.id,
        size: written
      });

      // Cleanup chunk temp file
      fs.unlink(tempPath, (err) => {
        if (err) console.error(`Failed to delete chunk temp file: ${tempPath}`, err);
      });
    }

    return {
      success: true,
      is_gpart: true,
      parts: uploadedParts
    };

  } catch (err) {
    // Rollback: delete uploaded parts if something fails
    console.error(`Upload split file failed: ${err.message}. Rolling back...`);
    for (const part of uploadedParts) {
      try {
        await deleteFile(part.account, part.file.id);
      } catch (delErr) {
        console.error(`Failed to delete part ${part.name} during rollback:`, delErr);
      }
    }
    return {
      success: false,
      error: `Gagal mengupload file: ${err.message}`
    };
  }
}

/**
 * Synchronize cache entries inside Database with Google Drive metadata.
 */
export async function syncAllAccountsCache(accounts) {
  const transaction = await sequelize.transaction();
  try {
    const accountIds = accounts.map(acc => acc.id);
    if (accountIds.length > 0) {
      // Delete old cache for these accounts
      await FileCache.destroy({
        where: { account_id: accountIds },
        transaction
      });
    }

    for (const account of accounts) {
      try {
        // 1. Fetch & update storage quota
        const quota = await getStorageQuota(account);
        account.quota_total = quota.total;
        account.quota_used = quota.used;
        if (quota.display_name) {
          account.display_name = quota.display_name;
        }
        await account.save({ transaction });

        // Get actual Google Drive Root ID to normalize into alias 'root'
        const { drive } = getDriveService(account);
        let rootId = null;
        try {
          const rootRes = await drive.files.get({ fileId: 'root', fields: 'id' });
          rootId = rootRes.data.id;
        } catch (e) {
          console.error(`Failed to retrieve root folder ID for account ${account.email}:`, e);
        }

        // 2. Fetch all files bulk metadata
        const driveFiles = await fetchAllFiles(account);
        
        // Prepare bulk insert entries
        const cacheEntries = driveFiles.map(f => {
          const parents = f.parents || [];
          let parentId = parents[0] || 'root';
          if (rootId && parentId === rootId) {
            parentId = 'root';
          }

          return {
            file_id: f.id,
            name: f.name,
            mime_type: f.mimeType,
            size: f.size ? parseFloat(f.size) : 0.0,
            modified_time: f.modifiedTime,
            parent_id: parentId,
            account_id: account.id,
            user_id: account.user_id,
            is_starred: f.starred ? 1 : 0,
            is_shared: f.shared ? 1 : 0
          };
        });

        if (cacheEntries.length > 0) {
          await FileCache.bulkCreate(cacheEntries, { transaction });
        }
      } catch (innerErr) {
        console.error(`Failed to sync files for account ${account.email}:`, innerErr);
      }
    }

    await transaction.commit();
    return true;
  } catch (e) {
    await transaction.rollback();
    console.error('Fatal error during cache synchronization:', e);
    return false;
  }
}

/**
 * Detect duplicate files in the cached database (files with matching name and size).
 */
export async function detectDuplicateFiles(userId) {
  // Query to find name and size that appear more than once
  const dupQuery = await FileCache.findAll({
    attributes: ['name', 'size'],
    where: {
      user_id: userId,
      size: { [Op.gt]: 0 },
      mime_type: { [Op.ne]: 'application/vnd.google-apps.folder' }
    },
    group: ['name', 'size'],
    having: sequelize.literal('count(id) > 1'),
    raw: true
  });

  const duplicates = [];
  for (const dup of dupQuery) {
    const matchingFiles = await FileCache.findAll({
      where: {
        user_id: userId,
        name: dup.name,
        size: dup.size
      }
    });

    const fileList = [];
    for (const f of matchingFiles) {
      const acc = await Account.findByPk(f.account_id);
      const email = acc ? acc.email : 'Unknown';

      fileList.push({
        id: f.file_id,
        name: f.name,
        size: f.size ? parseInt(f.size, 10) : 0,
        modifiedTime: f.modified_time,
        account_id: f.account_id,
        account_email: email
      });
    }

    duplicates.push({
      name: dup.name,
      size: dup.size ? parseInt(dup.size, 10) : 0,
      items: fileList
    });
  }

  return duplicates;
}

File: python-backend-backup/aggregator.py
"""
aggregator.py — Logika penggabungan virtual untuk beberapa akun Google Drive.
Termasuk: gabungan kuota, virtual directory, smart upload, caching sync, dan deteksi duplikat.
"""

from sqlalchemy import func
from drive_service import get_storage_quota, list_files, upload_file, fetch_all_files, get_drive_service
from models import FileCache, Account


def get_total_quota(accounts):
    """Hitung gabungan kuota penyimpanan dari semua akun."""
    total = 0
    used = 0
    account_quotas = []

    for account in accounts:
        try:
            quota = get_storage_quota(account)
            total += quota['total']
            used += quota['used']
            account_quotas.append({
                'account_id': account.id,
                'email': quota['email'],
                'display_name': quota['display_name'],
                'total': quota['total'],
                'used': quota['used'],
                'free': quota['total'] - quota['used']
            })
            # Update kuota di database
            account.quota_total = quota['total']
            account.quota_used = quota['used']
            if quota.get('display_name'):
                account.display_name = quota['display_name']
        except Exception as e:
            account_quotas.append({
                'account_id': account.id,
                'email': account.email,
                'display_name': account.display_name or account.email,
                'total': account.quota_total or 0,
                'used': account.quota_used or 0,
                'free': (account.quota_total or 0) - (account.quota_used or 0),
                'error': str(e)
            })

    return {
        'total': total,
        'used': used,
        'free': total - used,
        'accounts': account_quotas
    }


def get_merged_files(accounts, folder_id='root'):
    """Gabungkan daftar file dari semua akun menjadi satu daftar virtual."""
    merged = []
    for account in accounts:
        try:
            files = list_files(account, folder_id)
            merged.extend(files)
        except Exception as e:
            print(f"Error listing files for {account.email}: {e}")

    # Urutkan: folder dulu, lalu berdasarkan nama
    merged.sort(key=lambda x: (
        0 if x.get('mimeType') == 'application/vnd.google-apps.folder' else 1,
        x.get('name', '').lower()
    ))
    return merged


def merge_gpart_files(files_list):
    """
    Mengelompokkan file dengan pola .gpart.NNN menjadi satu file virtual terpadu.
    """
    import re
    pattern = re.compile(r'^(.*)\.gpart\.(\d+)$')
    groups = {}  # key: (parent_id, base_name) -> list of (part_num, file_dict)
    non_gparts = []

    for f in files_list:
        match = pattern.match(f['name'])
        if match:
            base_name = match.group(1)
            part_num = int(match.group(2))
            parent_id = f['parent_id']
            key = (parent_id, base_name)
            if key not in groups:
                groups[key] = []
            groups[key].append((part_num, f))
        else:
            non_gparts.append(f)

    for (parent_id, base_name), parts in groups.items():
        parts.sort(key=lambda x: x[0])
        total_size = sum(p[1]['size'] for p in parts)
        modified_times = [p[1]['modifiedTime'] for p in parts if p[1].get('modifiedTime')]
        latest_modified = max(modified_times) if modified_times else None
        
        is_starred = any(p[1].get('is_starred') for p in parts)
        is_shared = any(p[1].get('is_shared') for p in parts)
        
        first_part = parts[0][1]
        virtual_id = f"gpart:{parent_id}:{base_name}"
        
        virtual_file = {
            'id': virtual_id,
            'name': base_name,
            'mimeType': first_part.get('mimeType', 'application/octet-stream'),
            'size': int(total_size),
            'modifiedTime': latest_modified or first_part.get('modifiedTime'),
            'parent_id': parent_id,
            'account_id': first_part['account_id'],
            'is_starred': is_starred,
            'is_shared': is_shared,
            'is_gpart': True,
            'parts': [p[1] for p in parts]
        }
        non_gparts.append(virtual_file)

    return non_gparts


def smart_upload(accounts, file_stream, filename, mime_type, file_size, folder_id='root', progress_callback=None):
    """
    Upload file secara cerdas ke akun yang memiliki ruang paling banyak.
    Cek sisa kuota tiap akun. Jika tidak ada akun tunggal yang cukup,
    pecah file menjadi bagian-bagian dan upload ke beberapa akun.
    """
    import tempfile
    from drive_service import delete_file as drive_delete

    # Kumpulkan info ruang kosong per akun, urutkan dari yang paling banyak
    account_spaces = []
    for account in accounts:
        try:
            quota = get_storage_quota(account)
            free_space = quota['total'] - quota['used']
            account_spaces.append((account, free_space, quota['email']))
        except Exception as e:
            print(f"Error checking quota for {account.email}: {e}")
            continue

    # Urutkan berdasarkan ruang kosong terbesar
    account_spaces.sort(key=lambda x: x[1], reverse=True)

    # 1. Coba upload langsung ke akun tunggal yang muat
    for account, free_space, email in account_spaces:
        if free_space >= file_size:
            result = upload_file(account, file_stream, filename, mime_type, folder_id, progress_callback=progress_callback)
            return {
                'success': True,
                'file': result,
                'uploaded_to': email,
                'account_id': account.id
            }

    # 2. Jika tidak muat di satu akun, coba gunakan gabungan penyimpanan
    # Tetapkan batas aman (buffer) 10 MB per akun
    buffer_size = 10 * 1024 * 1024
    usable_spaces = []
    total_usable = 0
    for account, free_space, email in account_spaces:
        usable = max(0, free_space - buffer_size)
        if usable > 0:
            usable_spaces.append((account, usable, email))
            total_usable += usable

    if total_usable < file_size:
        return {
            'success': False,
            'error': 'Tidak ada akun dengan ruang penyimpanan yang cukup untuk file ini.'
        }

    # Tentukan alokasi ukuran part
    remaining_size = file_size
    allocations = []
    for account, usable, email in usable_spaces:
        if remaining_size <= 0:
            break
        allocated = min(remaining_size, usable)
        allocations.append((account, allocated, email))
        remaining_size -= allocated

    uploaded_parts = []
    try:
        for i, (account, allocated_size, email) in enumerate(allocations, 1):
            part_filename = f"{filename}.gpart.{i:03d}"
            
            # Buat file sementara untuk menampung part tersebut
            temp_fh = tempfile.TemporaryFile()
            bytes_written = 0
            while bytes_written < allocated_size:
                chunk_to_read = min(1024 * 1024, int(allocated_size - bytes_written))
                block = file_stream.read(chunk_to_read)
                if not block:
                    break
                temp_fh.write(block)
                bytes_written += len(block)

            if bytes_written < allocated_size:
                raise IOError(f"Unexpected EOF: expected {allocated_size} bytes, got {bytes_written}")

            temp_fh.seek(0)
            
            # Upload part ke Google Drive
            part_result = upload_file(
                account, temp_fh, part_filename, mime_type, folder_id,
                progress_callback=progress_callback
            )
            uploaded_parts.append((account, part_result, part_filename, bytes_written))
            temp_fh.close()

        return {
            'success': True,
            'is_gpart': True,
            'parts': [
                {
                    'file': res,
                    'name': name,
                    'uploaded_to': acc.email,
                    'account_id': acc.id,
                    'size': sz
                }
                for acc, res, name, sz in uploaded_parts
            ]
        }

    except Exception as e:
        # Rollback: Hapus part yang sudah sempat ter-upload jika terjadi error
        print(f"Upload split file gagal: {e}. Melakukan rollback...")
        for acc, res, name, sz in uploaded_parts:
            try:
                drive_delete(acc, res.get('id'))
            except Exception as del_err:
                print(f"Gagal menghapus part {name} saat rollback: {del_err}")
        return {
            'success': False,
            'error': f'Gagal mengupload file: {str(e)}'
        }



def sync_all_accounts_cache(db, accounts):
    """Ambil metadata terbaru dari semua akun Google Drive dan sinkronkan ke database lokal."""
    try:
        # Hapus cache file lama di database hanya untuk akun-akun ini
        account_ids = [acc.id for acc in accounts]
        if account_ids:
            db.query(FileCache).filter(FileCache.account_id.in_(account_ids)).delete(synchronize_session=False)
        
        for account in accounts:
            try:
                # 1. Ambil & perbarui kapasitas kuota akun terbaru
                quota = get_storage_quota(account)
                account.quota_total = quota['total']
                account.quota_used = quota['used']
                if quota.get('display_name'):
                    account.display_name = quota['display_name']
                
                # Ambil root ID riil dari Google Drive untuk dipetakan ke alias 'root'
                service, _ = get_drive_service(account)
                try:
                    root_id = service.files().get(fileId='root', fields='id').execute().get('id')
                except Exception as e:
                    print(f"Gagal mengambil root ID untuk {account.email}: {e}")
                    root_id = None
                
                # 2. Ambil seluruh file dari Google Drive secara massal
                drive_files = fetch_all_files(account)
                for f in drive_files:
                    parents = f.get('parents', [])
                    parent_id = parents[0] if parents else 'root'
                    
                    # Jika parent adalah root ID riil akun ini, petakan ke alias 'root'
                    if root_id and parent_id == root_id:
                        parent_id = 'root'
                        
                    # Simpan data file baru ke cache database
                    cache_item = FileCache(
                        file_id=f.get('id'),
                        name=f.get('name'),
                        mime_type=f.get('mimeType'),
                        size=float(f.get('size', 0)) if f.get('size') else 0.0,
                        modified_time=f.get('modifiedTime'),
                        parent_id=parent_id,
                        account_id=account.id,
                        user_id=account.user_id,
                        is_starred=1 if f.get('starred') else 0,
                        is_shared=1 if f.get('shared') else 0
                    )
                    db.add(cache_item)
            except Exception as inner_err:
                print(f"Gagal mensinkronisasikan file untuk {account.email}: {inner_err}")
                
        db.commit()
        return True
    except Exception as e:
        db.rollback()
        print(f"Error fatal saat menjalankan sinkronisasi cache: {e}")
        return False


def detect_duplicate_files(db, user_id):
    """
    Cari file duplikat (nama dan ukuran sama, size > 0) di database cache untuk user tertentu.
    Mengembalikan daftar file duplikat terkelompok.
    """
    # Cari nama dan ukuran file yang terduplikasi (bukan folder) untuk user aktif
    dup_query = db.query(FileCache.name, FileCache.size)\
        .filter(FileCache.user_id == user_id)\
        .filter(FileCache.size > 0)\
        .filter(FileCache.mime_type != 'application/vnd.google-apps.folder')\
        .group_by(FileCache.name, FileCache.size)\
        .having(func.count(FileCache.id) > 1)\
        .all()
        
    duplicates = []
    for name, size in dup_query:
        # Kueri semua instansi file yang memiliki kecocokan nama dan ukuran
        matching_files = db.query(FileCache).filter_by(user_id=user_id, name=name, size=size).all()
        
        file_list = []
        for f in matching_files:
            acc = db.query(Account).filter_by(id=f.account_id).first()
            email = acc.email if acc else 'Unknown'
            
            file_list.append({
                'id': f.file_id,
                'name': f.name,
                'size': int(f.size) if f.size else 0,
                'modifiedTime': f.modified_time,
                'account_id': f.account_id,
                'account_email': email
            })
            
        duplicates.append({
            'name': name,
            'size': int(size) if size else 0,
            'items': file_list
        })
        
    return duplicates
